packages/vm/src/environment.rs

```diff
@@ -60,31 +60,38 @@ impl GasState {
 
 /// A environment that provides access to the ContextData.
 /// The environment is clonable but clones access the same underlying data.
-pub struct Environment<S: Storage, Q: Querier> {
-    data: Arc<RwLock<ContextData<S, Q>>>,
+pub struct Environment<A: Api, S: Storage, Q: Querier> {
```

Comment:
Ah, nice, you did add the API to the environment

Reply:
Yeah, this is what you suggested. It turns out that using native functions, we can avoid all the runtime function signatures and get statically typed functions.